File: integration_tests/typescript/src/assertions.ts
import { deepStrictEqual } from 'assert';
import { unlinkSync, writeFileSync } from 'fs';

// The "omnifile" records the bytes of every value serialized by the exported functions below. It's
// used to validate that the TypeScript code generator encodes data identically to other code
// generators.
const omnifilePath = '/tmp/omnifile-typescript';

try {
  unlinkSync(omnifilePath);
} catch (_) {
  // Attempting to delete the file will fail if the file doesn't exist. This is harmless.
}

export function assertMatch<T, U>(
  size: (value: T) => number,
  serialize: (dataView: DataView, offset: number, value: T) => number,
  deserialize: (dataView: DataView) => U,
  actual: T,
  expected: U,
): void {
  console.log('Value to be serialized:', actual);

  const actualSize = size(actual);
  console.log('Expected size of the serialized value:', actualSize);

  const arrayBuffer = new ArrayBuffer(actualSize);
  const dataView = new DataView(arrayBuffer);

  const numBytesWritten = serialize(dataView, 0, actual);
  deepStrictEqual(numBytesWritten, actualSize);
  deepStrictEqual(arrayBuffer.byteLength, numBytesWritten);

  console.log('Bytes from serialization:', arrayBuffer);

  writeFileSync(omnifilePath, new Uint8Array(arrayBuffer), { flag: 'as' });

  console.log('Size of the serialized value:', arrayBuffer.byteLength);

  const replica = deserialize(dataView);

  console.log('Value deserialized from those bytes:', replica);

  deepStrictEqual(replica, expected);
}

export function assertRoundTrip<U, T extends U>(
  size: (value: T) => number,
  serialize: (dataView: DataView, offset: number, value: T) => number,
  deserialize: (dataView: DataView) => U,
  value: T,
): void {
  assertMatch(size, serialize, deserialize, value, value);
}

File: integration_tests/rust/src/assertions.rs
use {
    crate::types::{Deserialize, Serialize},
    std::{
        fmt::Debug,
        fs::{OpenOptions, remove_file},
        io::{self, Error, Write},
        mem::drop,
    },
};

const OMNIFILE_PATH: &str = "/tmp/omnifile-rust";

pub fn start() {
    drop(remove_file(OMNIFILE_PATH));
}

pub fn assert_match<T: Debug + Serialize, U: Debug + Deserialize>(
    actual: &T,
    expected: &U,
) -> io::Result<()> {
    println!("Message to be serialized: {actual:?}");

    let size = actual.size();
    println!("Expected size of the serialized value: {size:?}");

    let mut buffer = Vec::<u8>::new();
    actual.serialize(&mut buffer)?;
    println!("Bytes from serialization: {buffer:?}");
    println!("Size of the serialized value: {:?}", buffer.len());
    if buffer.len() != size {
        return Err(Error::other("Mismatch!"));
    }

    let mut file = OpenOptions::new()
        .create(true)
        .append(true)
        .open(OMNIFILE_PATH)
        .unwrap();
    file.write_all(&buffer).unwrap();

    let replica = U::deserialize(buffer.as_slice())?;
    println!("Message deserialized from those bytes: {replica:?}");

    if format!("{replica:?}") != format!("{expected:?}") {
        return Err(Error::other("Mismatch!"));
    }

    Ok(())
}

pub fn assert_round_trip<T: Debug + Serialize + Clone, U: Debug + Deserialize + From<T>>(
    value: &T,
) -> io::Result<()> {
    assert_match(value, &U::from(value.clone()))
}
